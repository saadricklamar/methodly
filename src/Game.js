import React, { Component } from 'react';
import './Game.scss';



class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      answer: '',
      currentCard: 0,
      currentQuestion: '',
      wrongCounter: 0,
      incorrectQuestions: []
    }

  }

  handleChange = (e) => {
    this.setState({answer: e.target.value})
  }

  matchAnswer = () => {
    this.state.currentCard++;
   if(this.props.stringMethodAnswers.length) {
      let result = this.props.stringMethodAnswers.find(methodAnswer => {
        return this.state.answer === methodAnswer 
    });
   if(this.state.answer === result) {
        this.rightAnswer();     
      } else {
        this.wrongAnswer();
      }
    this.setState({currentQuestion: this.props.stringMethodQuestions.pop()})
   } else if(this.props.mathMethodAnswers.length) {
      let result = this.props.mathMethodAnswers.find(methodAnswer => {
        return this.state.answer === methodAnswer 
    });
   if(this.state.answer === result) {
        this.rightAnswer();     
      } else {
        this.wrongAnswer();
      }
    this.setState({currentQuestion: this.props.mathMethodQuestions.pop()})
  }
  document.querySelector('#answer-input').value = '';
  this.setState({answer: ''});
}

  rightAnswer = () => {
   document.querySelector('.right-answer').style.display = 'block'
   setTimeout(this.hideRightAnswer, 2000)
  }

  hideRightAnswer= () => {
    document.querySelector('.right-answer').style.display = 'none'
  }


  wrongAnswer = () => {
    document.querySelector('.wrong-answer').style.display = 'block'
    setTimeout(this.hideWrong, 2000)
    this.state.wrongCounter++;
    if(this.state.wrongCounter === 1) {
      document.querySelector('#OneX').style.display = 'inline-block';
    } else if(this.state.wrongCounter === 2) {
      document.querySelector('#OneX').style.display = 'inline-block';
      document.querySelector('#TwoX').style.display = 'inline-block';
    } else if(this.state.wrongCounter === 3) {
      document.querySelector('#OneX').style.display = 'inline-block';
      document.querySelector('#TwoX').style.display = 'inline-block';
      document.querySelector('#ThreeX').style.display = 'inline-block';
    }
    this.state.incorrectQuestions.push(this.state.currentQuestion)
    this.saveToStorage();
  }

  hideWrong = () => {
    document.querySelector('.wrong-answer').style.display = 'none'
  }

  gameOver = () => {
    window.location.reload();
  }

   saveToStorage = () => {
      let stringifiedCards = JSON.stringify(this.state.incorrectQuestions);
      localStorage.setItem('saveToStudyLater', stringifiedCards);
  }
  
  studyIncorrectCards() {
      let localStorageCards = JSON.parse(localStorage.getItem('saveToStudyLater'))
      this.setState({currentQuestion: this.localStorageCards.pop()})
  } 

  render() { 
    console.log(this.state.currentCard)
    let question;
      if (this.state.currentCard === 0) {
         question = this.props.stringMethodQuestions[0] || this.props.mathMethodQuestions[0];
      } else {
        question = this.state.currentQuestion
      } 
      if (this.state.wrongCounter === 3) {
        document.querySelector('#answer-submit').disabled = true;
        question = `Oh no! You killed Socrates. You need more practice. Please click the restart button.`
      }
      if(this.state.currentCard === 10 && this.state.answer === '' || this.state.currentCard === 25 && this.state.answer === '') {
        document.querySelector('#answer-submit').disabled = true;
        question = 'Congrats on not killing Socrates! You are a master of these methods. Click restart to try another.'
      }
    return(
      <div className='welcome'>
        <p className='right-answer'>Correct Answer!</p>
        <p class='wrong-answer'>Wrong Answer!</p>
          <article className='card'>
            <h2 className='question'>{question}</h2> 
          </article>
          <input type='text' id='answer-input' placeholder='Enter Answer Here' onChange={this.handleChange}/>
          <input type='Submit' id='answer-submit' onClick={this.matchAnswer}/> 
          <input type='Submit' id='restart' value='Restart' onClick={this.gameOver}/> 
          <section className='wrong'>
            <i className='fas fa-times' id='OneX'></i>
            <i className='fas fa-times' id='TwoX'></i>
            <i className='fas fa-times' id='ThreeX'></i>
          </section>
      </div>
      )
  }




}


 export default Game;