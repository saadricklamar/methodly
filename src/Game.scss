@import './variables.scss';

.welcome {
  background: linear-gradient( rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.2) ), url(./images/Ancient.jpg) no-repeat center center fixed;
 -webkit-background-size: cover;
 -moz-background-size: cover;
 -o-background-size: cover;
 background-size: cover;
 display: flex;
 flex-direction: column;
 justify-content: center;
 align-items: center;
 height: 100vh;
 font-weight: 400;
 color: white;
}


.right-answer,
.wrong-answer {
  position: absolute;
  top: 200px;
  display: none;
  font-size: 50px;
  font-weight: bold;
  font-family: $primary-font;
  margin: 0;
}

.card {
  height: 200px;
  width: 400px;
  margin-top: 5px;
  border: 1px solid $primary-color;
  border-radius: 5px;
  background-color: $primary-background-color;
  .question {
    text-align: center;
    font-family: 'Bree Serif', serif;
    padding-top: 40px;
  }
}


#answer-submit,
#restart {
  width: 200px;
  height: 34px;
  margin-top: 30px;
  border-radius: 5px;
  background-color: $primary-background-color;
  font-family: $primary-font;
  font-weight: bold;
  font-size: 20px;
  color: $primary-color;
  cursor: $primary-pointer;
  &:hover {
    background-color: $primary-hover-color;
  }
}

#answer-input {
  width: 200px;
  height: 30px;
  margin-top: 50px;
  border-radius: 5px;
  font-size: 18px;
}

#OneX,
#TwoX,
#ThreeX {
    position: relative;
    right: 15px;
    color: red;
    font-size: 50px;
    display: none;
    margin-left: 30px;
    margin-top: 20px;
    margin-bottom: 0;
  }

