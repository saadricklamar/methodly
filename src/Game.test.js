import React from 'react';
import Game from './Game.js';
import { shallow } from 'enzyme';


const mockFunc = jest.fn();
const mockStringQuestions = ["Returns the character at the specified index (position)", "Returns the Unicode of the character at the specified index", "Joins two or more strings, and returns a new joined strings", "Checks whether a string ends with a specified string/characters", "Converts Unicode values to characters", "Checks whether a string contains the specified string/characters", "Returns the position of the first found occurence of a specified value in a string", "Returns the position of the last found occurence of a specified value in a string", "Compares two strings in the current locale", "Searches a string for a match against a regular expression, and returns the matches", "Searches a string for a specified value, or regula…expression, and returns the position of the match", "Searches a string for a specified value, or regula…expression, and returns the position of the match", "Extracts a part of a string and returns a new string", "Splits a string into an array of substrings", "Checks whether a string begins with specified characters", "Extracts the characters from a string, beginning a…on, and through the specified number of character", "Extracts the characters from a string, between two specified indices", "Converts a string to lowercase letters, according to the host's locale", "Converts a string to uppercase letters, according to the host's locale", "Converts a string to lowercase letters", "Returns the value of a String object", "Converts a string to uppercase letters", "Removes whitespace from both ends of a string", "Returns the primitive value of a String object"];
const mockStringAnswers = ["charAt()", "charCodeAt()", "concat()", "endsWith()", "fromCharCode()", "includes()", "indexOf()", "lastIndexOf()", "localCompare()", "match()", "replace()", "search()", "slice()", "split()", "startsWith()", "substr()", "substring()", "toLocaleLowerCase()", "toLocaleUpperCase()", "toLowerCase()", "toString()", "toUpperCase()", "trim()", "valueOf()"];
const mockMathQuestions = ["Returns a random number between 0 and 1", "Rounds x to the nearest integer", "Returns the value of x to the power of y", "Returns x, rounded downwards to the nearest integer", "Returns the number with the highest value", "Returns the number with the lowest value", "Returns the absolute value of x", "Returns x, rounded upwards to the nearest integer", "Returns the value of Ex"];
const mockMathAnswers = ["random()", "round(x)", "pow(x,y)", "floor(x)", "max(x,y,z,...,n)", "min(x,y,z,...,n)", "abs(x)", "ceil(x)", "exp(x)"];

describe('Game', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Game
                      stringMethodQuestions={mockStringQuestions}
                      stringMethodAnswers={mockStringAnswers}
                      mathMethodQuestions={mockMathQuestions}
                      mathMethodAnswers={mockMathAnswers} 
                      />)

  });
  it('Should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  it('Should have default states', () => {
    expect(wrapper.state()).toEqual ({
      answer: '',
      currentCard: 0,
      currentQuestion: '',
      wrongCounter: 0,
      incorrectQuestions: []
    });
  });
   it('Should set the state of answer', () => {
    expect(wrapper.state('answer')).toEqual('');
    const mockEvent = {target:{value: 'e'}}
    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state('answer')).toEqual('e');
  });
   it('mocks window.location.reload', () => {
    window.location.reload = jest.fn();
    expect(window.location.reload).not.toHaveBeenCalled();
    wrapper.instance().gameOver();
    expect(window.location.reload).toHaveBeenCalled();
    window.location.reload.mockRestore();
  });
  //  it('Should setState of currentQuestion', () => {
  //   const spyFunc = jest.fn();
  //   Object.defineProperty(global.document, '.wrong-answer', { value: spyFunc });
  //   expect(wrapper.state('currentQuestion')).toEqual('')
  //   wrapper.instance().matchAnswer(spyFunc);
  //   wrapper.setState({currentQuestion: mockStringQuestions.pop()})
  //   expect(wrapper.state('currentQuestion')).toEqual(mockStringQuestions.pop())
  // });
   
});
