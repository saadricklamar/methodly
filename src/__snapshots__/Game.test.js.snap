// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Should match the snapshot 1`] = `
<div
  className="welcome"
>
  <p
    className="right-answer"
  >
    Correct Answer!
  </p>
  <p
    class="wrong-answer"
  >
    Wrong Answer!
  </p>
  <article
    className="card"
  >
    <h2
      className="question"
    >
      Returns the character at the specified index (position)
    </h2>
  </article>
  <input
    id="answer-input"
    onChange={[Function]}
    placeholder="Enter Answer Here"
    type="text"
  />
  <input
    id="answer-submit"
    onClick={[Function]}
    type="Submit"
  />
  <input
    id="restart"
    onClick={[Function]}
    type="Submit"
    value="Restart"
  />
  <section
    className="wrong"
  >
    <i
      className="fas fa-times"
      id="OneX"
    />
    <i
      className="fas fa-times"
      id="TwoX"
    />
    <i
      className="fas fa-times"
      id="ThreeX"
    />
  </section>
</div>
`;
